src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php

```diff
@@ -0,0 +1,107 @@
+use Roave\BetterReflection\Identifier\Identifier;
+use Roave\BetterReflection\Identifier\IdentifierType;
+use Roave\BetterReflection\Reflection\Reflection;
+use Roave\BetterReflection\Reflection\ReflectionClass;
+use Roave\BetterReflection\Reflector\Reflector;
+use Roave\BetterReflection\SourceLocator\Ast\Strategy\NodeToReflection;
+use Roave\BetterReflection\SourceLocator\Exception\TwoAnonymousClassesOnSameLine;
+use Roave\BetterReflection\SourceLocator\Located\LocatedSource;
+use Roave\BetterReflection\Util\FileHelper;
+
+final class AnonymousClassObjectSourceLocator implements SourceLocator
+{
+    /**
+     * @var CoreReflectionClass
+     */
+    private $coreClassReflection;
+
+    /**
+     * @param object $anonymousClassObject
+     */
+    public function __construct($anonymousClassObject)
+    {
+        $this->coreClassReflection = new CoreReflectionClass($anonymousClassObject);
+    }
+
+    /**
+     * {@inheritDoc}
+     * @throws \Roave\BetterReflection\SourceLocator\Ast\Exception\ParseToAstFailure
+     */
+    public function locateIdentifier(Reflector $reflector, Identifier $identifier) : ?Reflection
+    {
+        return $this->getReflectionClass($reflector, $identifier->getType());
+    }
+
+    /**
+     * {@inheritDoc}
+     * @throws \Roave\BetterReflection\SourceLocator\Ast\Exception\ParseToAstFailure
+     */
+    public function locateIdentifiersByType(Reflector $reflector, IdentifierType $identifierType) : array
+    {
+        return array_filter([$this->getReflectionClass($reflector, $identifierType)]);
+    }
+
+    private function getReflectionClass(Reflector $reflector, IdentifierType $identifierType) : ?ReflectionClass
+    {
+        if (!$identifierType->isClass()) {
+            return null;
+        }
+
+        $fileName = FileHelper::normalizeWindowsPath($this->coreClassReflection->getFileName());
+
+        $ast = (new ParserFactory())->create(ParserFactory::PREFER_PHP7)->parse(file_get_contents($fileName));
+
+        $nodeVisitor = new class() extends NodeVisitorAbstract
+        {
+            /**
+             * @var Class_[]
+             */
+            private $anonymousClasses = [];
+
+            public function enterNode(Node $node) : void
+            {
+                if ($node instanceof Node\Stmt\Class_ && null === $node->name) {
+                    $this->anonymousClasses[] = $node;
+                }
+            }
+
+            /**
+             * @return Class_[]
+             */
+            public function getAnonymousClasses() : array
+            {
+                return $this->anonymousClasses;
+            }
+        };
+
+        $nodeTraverser = new NodeTraverser();
+        $nodeTraverser->addVisitor($nodeVisitor);
+        $nodeTraverser->traverse($ast);
+
+        /** @var Class_[] $anonymousClassNodesOnSameLine */
+        $anonymousClassNodesOnSameLine = array_values(array_filter($nodeVisitor->getAnonymousClasses(), function (Class_ $node) : bool {
```

Comment:
Can I suggest moving this into the anonmyous class? `getAnonymousClassesOnLine(int $lineNumber) : ?Class_` and let that API throw